src/character.cpp

```diff
@@ -10474,30 +10522,43 @@ bool Character::wearing_something_on( const bodypart_id &bp ) const
 
 bool Character::is_wearing_shoes( const side &which_side ) const
 {
-    bool left = true;
-    bool right = true;
-    if( which_side == side::LEFT || which_side == side::BOTH ) {
-        left = false;
-        for( const item &worn_item : worn ) {
-            if( worn_item.covers( body_part_foot_l ) && !worn_item.has_flag( flag_BELTED ) &&
-                !worn_item.has_flag( flag_PERSONAL ) && !worn_item.has_flag( flag_AURA ) &&
-                !worn_item.has_flag( flag_SEMITANGIBLE ) && !worn_item.has_flag( flag_SKINTIGHT ) ) {
-                left = true;
-                break;
+    /**
+     * nullopt - you have no foot
+     * true - you have a shoe
+     * false - you have no shoe but you have a foot
+     */
+
+    cata::optional<bool> left = cata::nullopt;
+    cata::optional<bool> right = cata::nullopt;
+    const bool left_side = which_side == side::LEFT || which_side == side::BOTH;
+    const bool right_side = which_side == side::RIGHT || which_side == side::BOTH;
+    const auto not_exempt = []( const item & worn_item ) {
+        return !worn_item.has_flag( flag_BELTED ) &&
+               !worn_item.has_flag( flag_PERSONAL ) && !worn_item.has_flag( flag_AURA ) &&
+               !worn_item.has_flag( flag_SEMITANGIBLE ) && !worn_item.has_flag( flag_SKINTIGHT );
+    };
+
+    for( const item &worn_item : worn ) {
+        for( const bodypart_id &part : get_all_body_parts() ) {
+            if( part->limb_type != body_part_type::type::foot ) {
+                continue;
             }
-        }
-    }
-    if( which_side == side::RIGHT || which_side == side::BOTH ) {
-        right = false;
-        for( const item &worn_item : worn ) {
-            if( worn_item.covers( body_part_foot_r ) && !worn_item.has_flag( flag_BELTED ) &&
-                !worn_item.has_flag( flag_PERSONAL ) && !worn_item.has_flag( flag_AURA ) &&
-                !worn_item.has_flag( flag_SEMITANGIBLE ) && !worn_item.has_flag( flag_SKINTIGHT ) ) {
-                right = true;
-                break;
+            if( !left_side && ( part->part_side == side::LEFT || part->part_side == side::BOTH ) ) {
+                left = worn_item.covers( part ) && not_exempt( worn_item );
+            }
+            if( !right_side && ( part->part_side == side::RIGHT || part->part_side == side::BOTH ) ) {
+                right = worn_item.covers( part ) && not_exempt( worn_item );
             }
         }
     }
+
+    if( right && !left.has_value() ) {
+        left = true;
+    }
+    if( left && !right.has_value() ) {
+        right = true;
+    }
+
     return ( left && right );
```

Comment:
I am not a hundred percent sure on this, if I remember correctly, if you put a `cata::optional` object in a boolean expression then it's actually evaluating whether this optional object has value or not. To evaluate the value inside one need to use `*` operator or `.value()` or `.value_or(default)`. Or maybe I'm wrong.

Reply:
You're right:
https://github.com/CleverRaven/Cataclysm-DDA/blob/dc4167363d5cabba4b20408e847775a2b836378e/src/optional.h#L119-L121

Reply:
Thank you for verifying this!

Reply:
oof. good catch.